
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>prism.extract module &#8212; prism  documentation</title>
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/language_data.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="prism.hmm module" href="prism.hmm.html" />
    <link rel="prev" title="prism.deconvolute module" href="prism.deconvolute.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="module-prism.extract">
<span id="prism-extract-module"></span><h1>prism.extract module<a class="headerlink" href="#module-prism.extract" title="Permalink to this headline">¶</a></h1>
<dl class="function">
<dt id="prism.extract.character_indices">
<code class="descclassname">prism.extract.</code><code class="descname">character_indices</code><span class="sig-paren">(</span><em>string</em>, <em>characters</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/prism/extract.html#character_indices"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#prism.extract.character_indices" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a sorted list of indices of characters in the string.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>string</strong> (<em>string</em>) – A string where the characters are expected to appear.</li>
<li><strong>characters</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.7)"><em>list</em></a>) – List of characters to return indices.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">A list of indices of string where the characters appear.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="prism.extract.discard_noninformative_reads">
<code class="descclassname">prism.extract.</code><code class="descname">discard_noninformative_reads</code><span class="sig-paren">(</span><em>cpgs_reads_dict</em>, <em>depth_cutoff</em>, <em>num_cpg_cutoff</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/prism/extract.html#discard_noninformative_reads"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#prism.extract.discard_noninformative_reads" title="Permalink to this definition">¶</a></dt>
<dd><p>Discards noninformative reads with low depth or few CpGs.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>cpgs_reads_dict</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.7)"><em>dict</em></a>) – Dictionary mapping a set of CpG coordinates to a group of reads.</li>
<li><strong>depth_cutoff</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.7)"><em>int</em></a>) – Minimum depth of a group of reads to be retained.</li>
<li><strong>num_cpg_cutoff</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.7)"><em>int</em></a>) – Mininum number of CpGs appearing in the reads to be retained.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">Retained (informative) CpG-Read group mapping dictionary.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="prism.extract.extend_region">
<code class="descclassname">prism.extract.</code><code class="descname">extend_region</code><span class="sig-paren">(</span><em>region</em>, <em>read</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/prism/extract.html#extend_region"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#prism.extract.extend_region" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the union of genomic regions covered by given region and read.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>region</strong> (<a class="reference internal" href="prism.util.html#prism.util.Region" title="prism.util.Region"><em>Region</em></a>) – Genomic region to be extended.</li>
<li><strong>read</strong> (<em>AlignedSegment</em>) – Read used for the extension.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">Extended genomic region, which is the union of region and read.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="prism.extract.get_cpg_coordinates">
<code class="descclassname">prism.extract.</code><code class="descname">get_cpg_coordinates</code><span class="sig-paren">(</span><em>read</em>, <em>paired=False</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/prism/extract.html#get_cpg_coordinates"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#prism.extract.get_cpg_coordinates" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns genomic coordinates of cytosines of CpGs in the read.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>read</strong> (<em>AlignedSegment</em>) – Read to examine for the location of CpGs.</li>
<li><strong>paired</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.7)"><em>bool</em></a>) – True if the sequencing library is paired, otherwise False.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">Absolute genomic coordinates of CpGs that appear in the read.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="prism.extract.has_overlap">
<code class="descclassname">prism.extract.</code><code class="descname">has_overlap</code><span class="sig-paren">(</span><em>region</em>, <em>read</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/prism/extract.html#has_overlap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#prism.extract.has_overlap" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if given region and read overlap, otherwise False.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>region</strong> (<a class="reference internal" href="prism.util.html#prism.util.Region" title="prism.util.Region"><em>Region</em></a>) – Genomic region to test overlap.</li>
<li><strong>read</strong> (<em>AlignedSegment</em>) – Read to test overlap.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">True if given region and read overlap, otherwise False.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="prism.extract.run">
<code class="descclassname">prism.extract.</code><code class="descname">run</code><span class="sig-paren">(</span><em>input_fp</em>, <em>output_fp</em>, <em>depth_cutoff</em>, <em>num_cpg_cutoff</em>, <em>prepend_chr</em>, <em>paired</em>, <em>verbose</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/prism/extract.html#run"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#prism.extract.run" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="prism.extract.save_met_file">
<code class="descclassname">prism.extract.</code><code class="descname">save_met_file</code><span class="sig-paren">(</span><em>handle</em>, <em>output_path</em>, <em>depth_cutoff</em>, <em>num_cpg_cutoff</em>, <em>paired</em>, <em>contigs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/prism/extract.html#save_met_file"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#prism.extract.save_met_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Read through the bam file and identify group of reads mapped to the same genomic region.
Group of reads with insufficient depth and insufficient number of CpGs will be discarded.
NOTE: BAM file should be sorted to guarantee proper generation of MET file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>handle</strong> (<em>AlignmentFile</em>) – pysam AlignmentFile object for BAM file.</li>
<li><strong>output_path</strong> (<em>string</em>) – Output MET file path.</li>
<li><strong>depth_cutoff</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.7)"><em>int</em></a>) – Minimum depth for a group of reads to be retained.</li>
<li><strong>num_cpg_cutoff</strong> – Minimum number of CpGs appearing in the reads to be retained.</li>
<li><strong>paired</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.7)"><em>bool</em></a>) – True if the sequencing library is paired, otherwise False.</li>
<li><strong>contigs</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.7)"><em>list</em></a>) – A list of contigs to examine.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">prism</a></h1>








<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="prism.html">prism package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="prism.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="prism.html#module-prism">Module contents</a></li>
</ul>
</li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="prism.html">prism package</a><ul>
      <li>Previous: <a href="prism.deconvolute.html" title="previous chapter">prism.deconvolute module</a></li>
      <li>Next: <a href="prism.hmm.html" title="next chapter">prism.hmm module</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019, Dohoon Lee.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.3</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../_sources/source/prism.extract.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>